### PR TITLE
Totem Boosts

### PR DESCRIPTION
**Pull request by ghoulslash**
## Description
- Adds totem boosts (from gen 7) to the battle engine.
- No config option is added because the user can just not set the totem boosts and it will never activate.
- Each stat can be raised/lowered by a unique value per battler.

Here is a GIF of a variety of stat changes occurring:
totem

### Mechanics
A `gTotemBoosts` EWRAM_DATA struct is added for each battler. It looks like:
```
struct TotemBoost {
    u8 stats;   //bitfield for each battle stat
    u8 statChanges[NUM_BATTLE_STATS - 1];    //highest bit is decrease
}; /* size = 8 */
```
- the `stats` element is a bitfield for each stat stage (7), with the last bit being used as a flag for the "totem flared to life" message.
- the `statChanges` element is 7 bytes (1 per stat stage), and is set to the number of stages to change by. The last bit, `0x80` is set to decrease the stat instead of raise
- A plethora of macros are added to the bottom of `asm/macros/event.inc` for +1 and +2 each independent stat. But the primary script macro, `settotemboost`, allows the user to customize their totem experience.

### CODE REVIEW

File: src/battle_script_commands.c [hunk 1]
```diff
@@ -8300,6 +8300,40 @@ static void Cmd_various(void)
...
+            gBattlescriptCurrInstr += 7;    //stats done, exit
+        }
+        else
+        {
+            for (i = 0; i < (NUM_BATTLE_STATS - 1); i++)
+            {
+                if (gTotemBoosts[gActiveBattler].stats & (1 << i))
+                {
+                    bool8 negative = (gTotemBoosts[gActiveBattler].statChanges[i] & 0x80) ? TRUE : FALSE;
+                    u8 change = gTotemBoosts[gActiveBattler].statChanges[i] & 0x7F;
```
Comment (ultima-soul) [inline src/battle_script_commands.c:8316]:
```suggestion
                    u8 change = gTotemBoosts[gActiveBattler].statChanges[i] & 0xF;
```
Is one solution to fixing an issue where numbers like -1 aren't properly being converted due to a cast from u8 to s8 in `ChangeStatBuff`.

**Review by ultima-soul — CHANGES_REQUESTED** [review]
Functions mostly fine aside from the negative numbers not working as expected.

One more thing to change is that the totem boost happens before abilities activate (like Drought etc.) from what I tested in USUM.

**Comment by ghoulslash** [comment]
Updates:
- allow negative script cmd values (instead of needing 0x80 bit set)
- Totem boosts occur before switch-in abilities:
totem

Ready for re-review

**Review by ultima-soul — DISMISSED** [review]

**Review by ultima-soul — APPROVED** [review]

**Review by ExpoSeed — CHANGES_REQUESTED** [review]
if `settotemboost` precedes `trainerbattle`, the game jumps to invalid address 101C101C

quick example to reproduce:
```arm
Route102_EventScript_Rick:: @ 81EC1B3
	settotemboost 0, -2, -2, -2, -2, -2, -2, -2
	trainerbattle_single TRAINER_RICK, Route102_Text_RickIntro, Route102_Text_RickDefeated
	msgbox Route102_Text_RickPostBattle, MSGBOX_AUTOCLOSE
	end
```

note that doing this:
```arm
Route102_EventScript_RouteSignOldale:: @ 81EC0EA
    settotemboost 0, -2, -2, -2, -2, -2, -2, -2
    msgbox Route102_Text_RouteSignOldale, MSGBOX_SIGN
    end
```
and then initiating a trainer battle did not lead to crash

**Comment by ghoulslash** [comment]
> if `settotemboost` precedes `trainerbattle`, the game jumps to invalid address 101C101C

@ExpoSeed 

This does not happen for me. You may need a `make clean` since I added EWRAM parameters

Proof:
totem-traienr

```
InsideOfTruck_EventScript_MovingBox::
	givemon SPECIES_AGGRON, 54, 0
	givemon SPECIES_BAGON, 15, 0
	settotemboost 0, -2, -2, -2, -2, -2, -2, -2
	trainerbattle_single TRAINER_RICK, intro, defeat
	end
intro:
	.string "hi$"
defeat:
	.string "bye$"
```

**Comment by ExpoSeed** [comment]
> This does not happen for me. You may need a `make clean` since I added EWRAM parameters

`make clean` did not affect anything

for what it's worth, i get crashes on this too. the trainer walking up completes, and then it crashes when the script starts
```arm
Route102_EventScript_Rick:: @ 81EC1B3
	settotemboost 0, -2, -2, -2, -2, -2, -2, -2
	end
```

In this situation, the msgbox never executes
```arm
Route102_EventScript_Rick:: @ 81EC1B3
	msgbox Route102_Text_RickPostBattle, MSGBOX_AUTOCLOSE
	settotemboost 0, -2, -2, -2, -2, -2, -2, -2
	end
```

This functions as expected
```arm
Route102_EventScript_RouteSignOldale:: @ 81EC0EA
	settotemboost 0, -2, -2, -2, -2, -2, -2, -2
	trainerbattle_single TRAINER_RICK, Route102_Text_RickIntro, Route102_Text_RickDefeated
	msgbox Route102_Text_RouteSignOldale, MSGBOX_SIGN
	end
```	
So now I think the bug has something to do with object events that use trainer sight, but i'm perplexed as to why you're not experiencing this issue

**Review by ExpoSeed — APPROVED** [review]
so apparently i'm just dumb

trainer_see requires `trainerbattle` to be the first script command, anything else crashes. it's a limitation of both gamefreak's scripting engine and their mental capacity

**Review by ultima-soul — APPROVED** [review]

**Pull request merged**